### ISSUE TITLE
db_bench cored, with exception of 'DATA SGL LENGTH INVALID'

### ISSUE THREAD

**Issue by ddwolf** [spdk/spdk#2489]
# Bug report

when I am issuing command `../rocksdb/db_bench --db=abc --benchmarks=fillrandom --num=10000000 --compression_type=none -spdk=rocksdb.json -spdk_bdev Nvme0n1 -spdk_cache_size=4096`, I got the Segmentation fault.

## Expected Behavior
db_bench can execute and return normally

## Current Behavior

```log
[2022-04-28 09:29:47.923855] Starting SPDK v22.01.1-pre git sha1 7fc39d249 / DPDK 21.11.0 initialization...
[2022-04-28 09:29:47.923955] [ DPDK EAL parameters: [2022-04-28 09:29:47.923968] rocksdb [2022-04-28 09:29:47.923977] --no-shconf [2022-04-28 09:29:47.923985] -c 0x1 [2022-04-28
 09:29:47.923992] --log-level=lib.eal:6 [2022-04-28 09:29:47.924000] --log-level=lib.cryptodev:5 [2022-04-28 09:29:47.924007] --log-level=user1:6 [2022-04-28 09:29:47.924015] --
iova-mode=pa [2022-04-28 09:29:47.924023] --base-virtaddr=0x200000000000 [2022-04-28 09:29:47.924031] --match-allocations [2022-04-28 09:29:47.924039] --file-prefix=spdk_pid9577
50 [2022-04-28 09:29:47.924047] ]
EAL: No free 2048 kB hugepages reported on node 1
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
[2022-04-28 09:29:47.998815] app.c: 601:spdk_app_start: *NOTICE*: Total cores available: 1
[2022-04-28 09:29:48.155126] app.c: 448:app_setup_trace: *NOTICE*: Tracepoint Group Mask 0x80 specified.
[2022-04-28 09:29:48.155200] app.c: 452:app_setup_trace: *NOTICE*: Use 'spdk_trace -s rocksdb -p 957750' to capture a snapshot of events at runtime.
[2022-04-28 09:29:48.155212] app.c: 454:app_setup_trace: *NOTICE*: Or copy /dev/shm/rocksdb_trace.pid957750 for offline analysis/debug.
[2022-04-28 09:29:48.155234] reactor.c: 943:reactor_run: *NOTICE*: Reactor started on core 0
[2022-04-28 09:29:48.156358] accel_engine.c: 510:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
RocksDB:    version 6.10
Date:       Thu Apr 28 09:30:51 2022
CPU:        56 * Intel(R) Xeon(R) Gold 5120 CPU @ 2.20GHz
CPUCache:   19712 KB
... finished 100 ops                              ^M... finished 200 ops                              ^M... finished 300 ops                              ^M... finished 400 ops
                             ^M... finished 500 ops                              ^M... finished 600 ops                              ^M... finished 700 ops
        ^M... finished 800 ops                              ^M... finished 900 ops                              ^M... finished 1000 ops                              ^M... finish
ed 1500 ops                              ^M... finished 2000 ops                              ^M... finished 2500 ops                              ^M... finished 3000 ops
                       ^M... finished 3500 ops                              ^M... finished 4000 ops                              ^M... finished 4500 ops
     ^M... finished 5000 ops                              ^M... finished 6000 ops                              ^M... finished 7000 ops                           
[2022-04-28 09:31:08.669494] nvme_qpair.c: 457:spdk_nvme_print_completion: *NOTICE*: DATA SGL LENGTH INVALID (00/0f) qid:1 cid:188 cdw0:0 sqhd:0041 p:1 m:0 dnr:0
[2022-04-28 09:31:08.669575] nvme_qpair.c: 274:nvme_io_qpair_print_command: *NOTICE*: WRITE sqid:1 cid:188 nsid:1 lba:14946816 len:512 SGL LAST SEGMENT ADDRESS 0x57d12fb048
[2022-04-28 09:31:08.669586] nvme_qpair.c: 457:spdk_nvme_print_completion: *NOTICE*: DATA SGL LENGTH INVALID (00/0f) qid:1 cid:188 cdw0:0 sqhd:0042 p:1 m:0 dnr:0
[2022-04-28 09:31:08.669669] nvme_qpair.c: 274:nvme_io_qpair_print_command: *NOTICE*: WRITE sqid:1 cid:188 nsid:1 lba:14946816 len:512 SGL LAST SEGMENT ADDRESS 0x57d12fb048
[2022-04-28 09:31:08.669680] nvme_qpair.c: 457:spdk_nvme_print_completion: *NOTICE*: DATA SGL LENGTH INVALID (00/0f) qid:1 cid:188 cdw0:0 sqhd:0043 p:1 m:0 dnr:0
[2022-04-28 09:31:08.669759] nvme_qpair.c: 274:nvme_io_qpair_print_command: *NOTICE*: WRITE sqid:1 cid:188 nsid:1 lba:14946816 len:512 SGL LAST SEGMENT ADDRESS 0x57d12fb048
[2022-04-28 09:31:08.669770] nvme_qpair.c: 457:spdk_nvme_print_completion: *NOTICE*: DATA SGL LENGTH INVALID (00/0f) qid:1 cid:188 cdw0:0 sqhd:0044 p:1 m:0 dnr:0
[2022-04-28 09:31:08.676611] nvme_qpair.c: 274:nvme_io_qpair_print_command: *NOTICE*: WRITE sqid:1 cid:191 nsid:1 lba:14950400 len:512 SGL LAST SEGMENT ADDRESS 0x57d12fe048
[2022-04-28 09:31:08.676626] nvme_qpair.c: 457:spdk_nvme_print_completion: *NOTICE*: DATA SGL LENGTH INVALID (00/0f) qid:1 cid:191 cdw0:0 sqhd:004c p:1 m:0 dnr:0
[2022-04-28 09:31:08.676706] nvme_qpair.c: 274:nvme_io_qpair_print_command: *NOTICE*: WRITE sqid:1 cid:191 nsid:1 lba:14950400 len:512 SGL LAST SEGMENT ADDRESS 0x57d12fe048
[2022-04-28 09:31:08.676717] nvme_qpair.c: 457:spdk_nvme_print_completion: *NOTICE*: DATA SGL LENGTH INVALID (00/0f) qid:1 cid:191 cdw0:0 sqhd:004d p:1 m:0 dnr:0
[2022-04-28 09:31:08.676800] nvme_qpair.c: 274:nvme_io_qpair_print_command: *NOTICE*: WRITE sqid:1 cid:191 nsid:1 lba:14950400 len:512 SGL LAST SEGMENT ADDRESS 0x57d12fe048
[2022-04-28 09:31:08.676815] nvme_qpair.c: 457:spdk_nvme_print_completion: *NOTICE*: DATA SGL LENGTH INVALID (00/0f) qid:1 cid:191 cdw0:0 sqhd:004e p:1 m:0 dnr:0
[2022-04-28 09:31:08.676894] nvme_qpair.c: 274:nvme_io_qpair_print_command: *NOTICE*: WRITE sqid:1 cid:191 nsid:1 lba:14950400 len:512 SGL LAST SEGMENT ADDRESS 0x57d12fe048
[2022-04-28 09:31:08.676905] nvme_qpair.c: 457:spdk_nvme_print_completion: *NOTICE*: DATA SGL LENGTH INVALID (00/0f) qid:1 cid:191 cdw0:0 sqhd:004f p:1 m:0 dnr:0
put error: Corruption: block checksum mismatch: expected 1662653421, got 2324967102  in abc/000053.sst offset 53042022 size 3999
using bdev Nvme0n1
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
Keys:       16 bytes each
Values:     100 bytes each (50 bytes after compression)
Entries:    10000000
Prefix:    0 bytes
Keys per prefix:    0
RawSize:    1106.3 MB (estimated)
FileSize:   629.4 MB (estimated)
Write rate: 0 bytes/second
Read rate: 0 ops/second
Compression: NoCompression
Compression sampling rate: 0
Memtablerep: skip_list
Perf Level: 1
------------------------------------------------
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
DB path: [abc]
Received signal 11 (Segmentation fault)
#0   ../rocksdb/db_bench() [0x9042a0]
```

## Steps to Reproduce

1. compile spdk
```bash
./configure --prefix=/path/to/spdk --with-fuse
make
```
2. compile db_bench
```bash
make db_bench SPDK_DIR=/path/to/spdk/ EXTRA_CFLAGS=/path/to/spdk//include EXTRA_CXXFLAGS=/path/to/spdk/include -j
```
3. execute db_bench
```bash
 ../rocksdb/db_bench --db=abc --benchmarks=fillrandom --num=10000000 --compression_type=none -spdk=rocksdb.json -spdk_bdev Nvme0n1 -spdk_cache_size=4096 &> a
```
4. here is the rocksdb.json
```json

{
  "subsystems": [
    {
      "subsystem": "bdev",
      "config": [
        {
          "method": "bdev_nvme_attach_controller",
          "params":
          {
            "trtype": "PCIe",
            "name":"Nvme0",
            "traddr":"0000:d8:00.0"
          }
        }]
    }]
}
```

## Context (Environment including OS version, SPDK version, etc.)

OS: eulerosv2p9
spdk: v22.01.x
rocksdb: spdk-v6.10.2

**Comment by Comphix** [comment]
Looks like some chuecksum mismatch found

[2022-04-28 09:31:08.676905] nvme_qpair.c: 457:spdk_nvme_print_completion: *NOTICE*: DATA SGL LENGTH INVALID (00/0f) qid:1 cid:191 cdw0:0 sqhd:004f p:1 m:0 dnr:0
put error: Corruption: block checksum mismatch: expected 1662653421, got 2324967102  in abc/000053.sst offset 53042022 size 3999
using bdev Nvme0n1

**Comment by Comphix** [comment]
Hailiang, please first have a reproduction on latest SPDK code. Thanks, Gang.

**Comment by changpe1** [comment]
I don't think we have a SSD drive can support 256KiB MDTS and hardware SGL feature, we can have a try with P5500 serial SSD, otherwise it doesn't make sense to us.

**Comment by wanghailiangx** [comment]
OK, I will have a look with P5500.

**Comment by changpe1** [comment]
@ddwolf can you share your SSD models?

**Comment by ddwolf** [comment]
@changpe1 
```bash
# lspci | grep -i nvme
d8:00.0 Non-Volatile memory controller: Huawei Technologies Co., Ltd. ES3000 V3 NVMe PCIe SSD (rev 45)
```

**Comment by ddwolf** [comment]
After I turned down the --num parameter to 100, and turned up the hugepage to a much bigger number, I got the Hugepage warning away and the db_bench works fine.
` ../rocksdb/db_bench --db=abc --benchmarks=fillrandom --num=100 --compression_type=none -spdk=rocksdb.json -spdk_bdev Nvme0n1 -spdk_cache_size=4096 &> a
`

**Comment by changpe1** [comment]
@ddwolf you may refer to https://github.com/spdk/spdk/issues/2377, Huawei SSDs has an issue when SGL is dword alinged.

**Comment by jimharris** [comment]
> After I turned down the --num parameter to 100, and turned up the hugepage to a much bigger number, I got the Hugepage warning away and the db_bench works fine.
> `../rocksdb/db_bench --db=abc --benchmarks=fillrandom --num=100 --compression_type=none -spdk=rocksdb.json -spdk_bdev Nvme0n1 -spdk_cache_size=4096 &> a`

Probably the SGL issue only happens with very specific buffer size/alignments.  So reducing the number of operations from 10M to 100 means it is much less likely to hit the SGL issue.

Can you try disabling setting of the SPDK_NVME_CTRLR_SGL_SUPPORTED flag in lib/nvme/nvme_ctrlr.c?  Just as a local patch - this will disable use of SGL and use PRP instead.  I think this may pass on your system.

**Comment by changpe1** [comment]
@ddwolf  can you try with https://review.spdk.io/gerrit/c/spdk/spdk/+/12405 to see this issue is still exist or not ?

**Comment by ddwolf** [comment]
@jimharris @changpe1 after apllying patch 12405, everything goes well to the end. Thanks you.

**Comment by changpe1** [comment]
> @jimharris @changpe1 after apllying patch 12405, everything goes well to the end. Thank you.
> 
> And another question, should disabling SGL would cause an inefficiency?

It should be very trivial, especially when data buffer is contiguous, but you can use SPDK examples/perf tool to have a test with or without this patch.

**Comment by wanghailiangx** [comment]
> Hailiang, please first have a reproduction on latest SPDK code. Thanks, Gang.
I have verified the user's steps with P5800(have no P5500) and P3700, can't reproduce the issue.
OS: Fedora 32
spdk: the latest
rocksdb: spdk-v6.10.2

**Comment by changpe1** [comment]
> > Hailiang, please first have a reproduction on latest SPDK code. Thanks, Gang.
> > I have verified the user's steps with P5800(have no P5500) and P3700, they does not have this issue.
> > OS: Fedora 32
> > spdk: the latest
> > rocksdb: spdk-v6.10.2

@wanghailiangx   Only P5500 can be used for this test case.

**Comment by ddwolf** [comment]
------------------------------------
With the patch
------------------------------------
```bash
[spdk]# ./build/examples/perf -q 128 -o 4096 -w randread -r 'trtype:PCIe traddr:0000:d8:00.0' -t 300
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
Initializing NVMe Controllers
Attached to NVMe Controller at 0000:d8:00.0 [19e5:0123]
Associating PCIE (0000:d8:00.0) NSID 1 with lcore 0
Initialization complete. Launching workers.
========================================================
                                                                           Latency(us)
Device Information                     :       IOPS      MiB/s    Average        min        max
PCIE (0000:d8:00.0) NSID 1 from core  0:  735246.33    2872.06     174.07       8.36    5369.91
========================================================
Total                                  :  735246.33    2872.06     174.07       8.36    5369.91
```
------------------------------------
Without the patch
------------------------------------
```bash
========================================================
                                                                           Latency(us)
Device Information                     :       IOPS      MiB/s    Average        min        max
PCIE (0000:d8:00.0) NSID 1 from core  0:  736772.32    2878.02     173.71       8.70    5126.34
========================================================
Total                                  :  736772.32    2878.02     173.71       8.70    5126.34

```
@changpe1 looks that the difference is tiny but obvious, min latency tends to be worse, right?

**Comment by wanghailiangx** [comment]
@ddwolf Once or twice the data is accidental. I suggest you test it several times to see the average difference.
